### PR TITLE
DeepseekV4: fix seq_rm

### LINKED ISSUE #25566: Misc. bug: server - deepseek4 slot KV never reclaimed on release; fails with "Context size has been exceeded" when cumulative prompts exceed n_ctx
### Name and Version

c749cb041 (9966), CPU backend, Ubuntu 24.04

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Command line

```shell
llama-server -m <DeepSeek-V4-Flash UD-Q4_K_XL gguf> -ngl 0 -fa 1 --jinja --reasoning-budget 0 --parallel 1 --ctx-size 32768 --threads 32
```

### Problem description & steps to reproduce

Running `llama-server` with a `deepseek4` model, KV cache consumed by COMPLETED requests
is apparently never reclaimed when the slot is released. The server hard-fails mid-
generation with `Context size has been exceeded` once the **sum of prompt sizes since
startup** exceeds the KV pool — even with a single slot and every prior request finished
and released normally. Reproduced at three different configs. May share a root cause with
the model's hybrid-attention KV handling; possibly related to #25468 (ctx-proportional
buffers) but the symptom here is cross-request accumulation, not per-request size.

## Environment

- build: c749cb041 (9966), CPU backend (GGML_BLAS=OFF), Ubuntu 24.04
- hardware: AMD EPYC 9V45 (Turin), 32 cores / 512 GB
- model: unsloth/DeepSeek-V4-Flash-GGUF UD-Q4_K_XL
- server: `llama-server -ngl 0 -fa 1 --jinja --reasoning-budget 0 --temp 0.1` + configs below
- client: OpenAI-compat /v1/chat/completions, strictly sequential requests (one at a time),
  each ~256 predicted tokens, prompts growing ~280 tokens per request (agent-loop pattern)

## Reproductions (3 configs, same failure)

| Config | Requests completed before failure | Σ prompt tokens at failure |
|---|---|---|
| `--parallel 2 --ctx-size 16384` (8K/slot) | 6 | ~7.3K + generations ≈ pool |
| `--parallel 1 --ctx-size 32768` | 5 | ~30K ≈ pool |
| `--parallel 1 --ctx-size 32768` (fresh restart, rerun) | ~9 | ≈ pool again |

Workaround that confirms the arithmetic: `--ctx-size 131072` completed 14 further requests
(Σ ≈ 110K < 128K) without failure.

## Failure signature

```
W srv decode: failed to find free space in the KV cache, retrying with smaller batch size, off = 0, n_batch = 2048, ret = 1
... (cascade n_batch 2048 -> 1) ...
E srv decode: Context size has been exceeded. off = 0, n_batch = 1, ret = 1
E srv send_error: task id = 1553, error: Context size has been exceeded.
I slot release: id 1 | task 1553 | stop processing: n_tokens = 1952, truncated = 0
```

Note the failing request itself is small (n_tokens 1952 on an 8K slot); prior releases all
logged normally. Failure occurs mid-generation, returns HTTP 500 to the client.

## Expected

Slot release returns its KV to the pool; a single-slot server should sustain unlimited
sequential requests whose individual size fits the slot.

Can provide full server logs / rerun with instrumentation.

### Relevant log output

<details>
<summary>Logs</summary>
W srv decode: failed to find free space in the KV cache, retrying with smaller batch size, off = 0, n_batch = 2048, ret = 1
... (cascade n_batch 2048 -> 1) ...
E srv decode: Context size has been exceeded. off = 0, n_batch = 1, ret = 1
E srv send_error: task id = 1553, error: Context size has been exceeded.
I slot release: id 1 | task 1553 | stop processing: n_tokens = 1952, truncated = 0

```console

```
</details>

### PR DESCRIPTION
**Pull request by am17an**
## Overview

Fixes #25566. This PR adds clears the compressed caches till the checkpoint as well after a server checkpoint restore.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure:

### CODE REVIEW

File: src/llama-kv-cache-dsv4.cpp [hunk 1]
```diff
@@ -1174,14 +1174,19 @@ bool llama_kv_cache_dsv4::seq_rm(llama_seq_id seq_id, llama_pos p0, llama_pos p1
...
+            return false;
         }
 
-        return false;
+        bool res = true;
+
+        res = res & kv_raw->seq_rm(seq_id, p0, -1);
+        res = res & kv_csa->seq_rm(seq_id, p0/DSV4_CSA_RATIO, -1);
+        res = res & kv_hca->seq_rm(seq_id, p0/DSV4_HCA_RATIO, -1);
+        res = res & kv_lid->seq_rm(seq_id, p0/DSV4_CSA_RATIO, -1);
```
Comment (ggerganov) [inline src/llama-kv-cache-dsv4.cpp:1225]: AFAIU, this logic does not work correctly for `p0 % DSV4_*_RATIO != 0`? If yes, add TODO

Comment (am17an) [reply src/llama-kv-cache-dsv4.cpp:1225]: It will work since the compressed state holds on the rest of the tokens, these caches only contain the committed tokens (post compression)

File: src/llama-kv-cache-dsv4.cpp
Comment (ggerganov) [inline src/llama-kv-cache-dsv4.cpp:1250]: Do we need to copy the rest of the caches too? Also the compressed states?

Comment (am17an) [reply src/llama-kv-cache-dsv4.cpp:1250]: The entire `seq_cp` machinery was missing for those states, added

Comment (ggerganov) [reply src/llama-kv-cache-dsv4.cpp:1250]: The `test-save-load-state` should have caught that - make sure you run it when making changes to the KV cache logic.

Comment (am17an) [reply src/llama-kv-cache-dsv4.cpp:1250]: Yes it currently fails early and exits for the test case I discussed earlier. Need to make changes to the test

File: src/llama-kv-cache-dsv4.cpp [hunk 2]
```diff
@@ -1720,6 +1769,16 @@ bool llama_kv_cache_dsv4_context::apply() {
...
+    if (ctx_csa_mem) {
+        res = res & ctx_csa_mem->apply();
+        res = res & ctx_hca_mem->apply();
+        res = res & ctx_lid_mem->apply();
+
+        csa_state->apply_copies();
+        hca_state->apply_copies();
+        lid_state->apply_copies();
+    }
+
```
Comment (ggerganov) [inline src/llama-kv-cache-dsv4.cpp:1792]: The updates of the memory modules (such as stream copies) are handled through "update contexts". These are initialized with `llama_memory_i::init_update()`:

https://github.com/ggml-org/llama.cpp/blob/e920c523e3b8a0163fe498af5bf90df35ff51d25/src/llama-memory.h#L96-L98

They should not contain any ubatches and just the stream copy informations:

https://github.com/ggml-org/llama.cpp/blob/e920c523e3b8a0163fe498af5bf90df35ff51d25/src/llama-kv-cache.cpp#L2540-L2549

File: src/llama-kv-cache-dsv4.cpp
Comment (ggerganov) [reply src/llama-kv-cache-dsv4.cpp:1250]: There is currently an error that does not seem related to the logits matching that we discussed:

```bash
make -j && ./bin/test-save-load-state -m DeepSeek-V4-Flash-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2-imatrix.gguf

0.03.896.416 I main: no prompt provided, generating 100 (n_batch) random tokens
0.03.896.425 I main: the input prompt is 100 tokens
0.04.619.391 W resolve_fused_ops: layer 2 is assigned to device MTL0 but Lightning Indexer is assigned to device CPU (usually due to missing support)
0.04.619.396 W resolve_fused_ops: Lightning Indexer not supported, set to disabled
0.07.105.955 I cmn  common_promp: saved session before last token to dump_state.bin, n_new = 100

=== Test 1: baseline ===
20274 67413 33587 58425 79374 2573 3171 20759 428 499 16 10195 1951 29326 477 95 
0.09.324.848 W resolve_fused_ops: layer 2 is assigned to device MTL0 but Lightning Indexer is assigned to device CPU (usually due to missing support)
0.09.324.852 W resolve_fused_ops: Lightning Indexer not supported, set to disabled

=== Test 2: sequence removal isolation ===
PASS
0.11.912.031 W resolve_fused_ops: layer 2 is assigned to device MTL0 but Lightning Indexer is assigned to device CPU (usually due to missing support)
0.11.912.039 W resolve_fused_ops: Lightning Indexer not supported, set to disabled

=== Test 3: state load ===
20274 67413 33587 58425 79374 2573 3171 20759 428 499 16 10195 1951 29326 477 95 
PASS
0.16.853.990 W resolve_fused_ops: layer 2 is assigned to device MTL0 but Lightning Indexer is assigned to device CPU (usually due to missing support)
0.16.853.999 W resolve_fused_ops: Lightning Indexer not supported, set to disabled

=== Test 4: seq copy (host) ===
0.17.303.079 E llama_state_load_file: error loading session file: n_stream mismatch
0.17.303.204 E 
test_seq_cp_host: failed to load state
```

Do you repro on your end?

Comment (am17an) [reply src/llama-kv-cache-dsv4.cpp:1250]: Yes since it does not properly support unified cache. `-np 2` will pass

File: src/llama-kv-cache-dsv4.cpp (hunk 2 above)
Comment (am17an) [reply src/llama-kv-cache-dsv4.cpp:1792]: @ggerganov not sure I follow, the `apply_copies` function only contains the stream copies?

Comment (ggerganov) [reply src/llama-kv-cache-dsv4.cpp:1792]: If I follow the logic correctly, the problem is that the stream copies are performed together with the ubatch processing. The convention is to perform the stream copies before that, during `llama_context::memory_update()`. To do that, the stream copies have to be passed to an separate "update context" that is initialized with `init_update()` and does not contain any ubatches.

**Review by ggerganov — APPROVED** [review]

**Review by fairydreaming — APPROVED** [review]
